Mathlib/LinearAlgebra/Multilinear/Basic.lean

```diff
@@ -760,8 +763,151 @@ theorem domDomCongr_eq_iff (σ : ι₁ ≃ ι₂) (f g : MultilinearMap R (fun _
 
 end
 
+
+section Derivative
+
+/-! The goal of this part is, for `f` a multilinear map in finitely many variables indexed by `ι`,
+to develop `f(x + y)` as a finite formal multilinear series in `y`. The term of degree `n` of
+this formal multilinear series will be a sum over finsets `s` of `ι` of cardinality `n` of a
+multilinear map indexed by `s`, evaluated at `fun (_ : s) => y`; we use a linear order on
+`ι` to identify all such `s` to `Fin n`. We also give a more direct definition of the term
+of degree `1` in `f.linearDeriv`, which doesn't require a linear order on `ι`, and prove the
+equivalence of the two definitions.-/
+
+lemma domDomRestrict_aux [DecidableEq ι] (s : Set ι) [(i : ι) → Decidable (i ∈ s)]
+    [DecidableEq s] (x : (i : s.compl)→ M₁ i) (z : (i : s) → M₁ i) (i : s)
+    (c : M₁ i) : (fun j ↦ if h : j ∈ s then Function.update z i c ⟨j, h⟩ else x ⟨j, h⟩) =
+    Function.update (fun j => if h : j ∈ s then z ⟨j, h⟩ else x ⟨j, h⟩) i c := by
```

Comment:
@YaelDillies, seems this function is appearing again; perhaps we really should have `Function.piecewise` or `Subtype.piecewise`?

Reply:
Should that replace `Set.piecewise`?

Reply:
I think we can keep `Set.piecewise` for the non-dependent version

Reply:
That's quite confusing. `Set.piecewise` *is* dependent, but not on the set membership.